### PR TITLE
tidb: reset transaction when session retry statement in history meets error

### PR DESCRIPTION
**Pull request by tiancaiamao**
It's not easy to construct the testcase for this change, I'll refactor mockTikv and add it later.
PTAL @shenli @disksing

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -384,6 +384,15 @@ func (s *session) retry(maxCnt int, infoSchemaChanged bool) error {
...
 			_, err = st.Exec(s)
 			if err != nil {
+				// begin
+				// insert xxx
+				// update xxx
+				// commit
+				//
+				// update xxx need to read data first, if it meets error, we should reset txn before next retry.
+				// otherwise, we'll get duplicate key error during next retry because insert operation has already done.
+				s.txn = nil
```
Comment (shenli) [inline session.go]: Can we clean txn after line 415? Or the log will print a nil for s.txn.

File: session.go [hunk 2]
```diff
@@ -384,6 +384,15 @@ func (s *session) retry(maxCnt int, infoSchemaChanged bool) error {
 			s.sessionVars.StmtCtx.ResetForRetry()
 			_, err = st.Exec(s)
 			if err != nil {
+				// begin
+				// insert xxx
+				// update xxx
+				// commit
+				//
+				// update xxx need to read data first, if it meets error, we should reset txn before next retry.
+				// otherwise, we'll get duplicate key error during next retry because insert operation has already done.
```
Comment (disksing) [inline session.go]: I think the example in comment is too specific.

**Comment by disksing** [comment]
LGTM.

**Comment by shenli** [comment]
LGTM

**Pull request merged**